pkg/codegen/testing/test/testdata/functions-secrets/nodejs/tsconfig.json

```diff
@@ -10,6 +11,9 @@
         "experimentalDecorators": true,
         "noFallthroughCasesInSwitch": true,
         "forceConsistentCasingInFileNames": true,
+		"paths": {
+			"@/*": ["*"]
+		},
```

Comment:
The tsconfig changes seem like no-ops. Are they? 

`baseUrl` and `paths` don't appear used, as I don't see any absolute paths, and they introduce the potential for name collisions between absolute paths and the names of dependencies. 

As we're generating all of these files, we shouldn't need path mangling: we can generate correct paths from the get-go:


baseURL: `import * as x from "hello"` could resolve to `./hello` or `node_modules/hello`

paths: `import * as x from "@/foo"` doesn't appear used.

Reply:
@AaronFriel It's not a no-op. It's used [here](https://github.com/pulumi/pulumi/pull/10831/files#diff-f7917202f932e91c5a03a811977f165efebbc5c006159c0a8865e3a93713048cR1655).

TL;DR: The mangling will be removed in https://github.com/pulumi/pulumi/issues/8613 and having absolute paths makes that much easier IMO.

> As we're generating all of these files, we shouldn't need path mangling: we can generate correct paths, rather than relying on aliasing/rewrite rules in tsconfig.

I assume you're referring to [this snippet](https://github.com/pulumi/pulumi/pull/10831/files#diff-f7917202f932e91c5a03a811977f165efebbc5c006159c0a8865e3a93713048cR1647-R1659), which translates paths relative to `/types` to a top-level path.

You're right, it is technically possible to track the correct relative path and use that instead. I would have used that approach but it became probability invasive. Here's why:

I found that the existing logic for generating relative paths incorrect. `mod.getRelativePath()` operates using the `mod.mod` field, which is always a single path, so the function will always return `..` no matter what mod it's operating on. Further, I had a _rough_ time adapting the logic to the nested types directories. Many functions assume that the code generated is at most two levels removed from the parent at any time, which is not true after this PR. Lots of high level functions call `getRelativePath` deeply in their callstack.
One workaround is to adopt a recursive approach where we carry the context of how deep we are between function calls. In general, this would need to adjust some number of functions to accept a `parentDir` parameter to track how deep we are in the file hierarchy (or, since we know the depth of the filesystem is <= 2, do something less DRY). This parameter needs to be threaded to a number of existing methods that don't already have it, and any calls to those methods need to supply a default value of some kind. 

I actually implemented this on `intoIOFiles` before realizing I could use `paths` (it's still in there now). The knock-on effects of threading this through `intoIOFiles`:

1. Must add the parameter to `genOwnedTypes`.
2. ...which means I have to add the parameter to `sdk.ImportsWithPath`
3. ...which means threading it through `getRelativePath` because the original logic was no longer valid. 
4. Same with `genNestedTypes` because it is corecursive with `intoIOFiles`.
5. Same with `genIndexFile`

I think, ultimately, the architecture using an absolute path will look similar to what you have in mind here. As part of https://github.com/pulumi/pulumi/issues/8613 we need to build a map between type decls and their defn dependencies. If we use relative paths, then we have to recompute the map for each subdirectory. If we use absolute paths, we only need to compute it once (or once per mod context, depending on which end of the elephant you stare at).

I actually think mapping the package root to `@` via `paths` can dramatically simplify our code, both by letting us refactor existing code and also making future changes easier.
For existing code, this lets us rip out `getRelativePath`, which as I mentioned is vacuously correct. For future code, it enabled a cleaner approach to implementing https://github.com/pulumi/pulumi/issues/8613.

Aside: I also think for most of our imports abs paths will be easier to read in the generated code than relative paths. Most of the imports are `@/s3` or `@/ec2`. With relative paths, you can have both `../../s3` and `../s3`, and they point to the same file.

Reply:
I hear you, though I looked through (twice now) all of the files generated by tests and no lines change match either of these patterns. If the PR introduces functionality not tested, I'm not sure how I'm to evaluate it - would this appear if I used this branch to run sdkgen against a provider?

Reply:
FWIW, if I use a tool instead of eyeballing it: `git diff origin/master..HEAD | rg 'from "@'`  and likewise `from "\w+`.

Reply:
I don't disagree with using absolute paths if it simplifies codegen & your case with @t0yv0 was compelling, apologies if my comment lead you to believe I'm opposed in general.

Reply:
I thought you were asking if it was used in the generator anywhere, I gotcha now! :) 

> I looked through (twice now) all of the files generated by tests and no lines change match either of these patterns.

Now that you mention it, I think you're right that the codegen tests don't exercise that code path. I introduced it as a fix failing downstream codegen CI jobs. Those jobs exercise the codepath in `pulumi-aws` at the least (not sure where else).

It occurs whenever an input or output type needs to reference a resource type.

Here's an example from `pulumi-aws/types/s3/output.ts`, which imports `pulumi-aws/s3`:
<img width="839" alt="Screen Shot 2022-11-09 at 6 28 20 PM" src="https://user-images.githubusercontent.com/4955075/200963934-29fff3a1-4dce-474b-a3ec-cdc61baa17f3.png">

Do you want me to add a codegen test for this case?

> apologies if my comment lead you to believe I'm opposed in general.

No need to apologize! I appreciate the thorough code review. I wanted to make my case for the absolute imports anyway because I want to get feedback from yinz guys on whether its a good idea or not :) This is precisely what I want :)

Reply:
Yes, definitely to adding or modifying an existing codegen test to help us catch a regression.

And a nit on the tsconfig file format: it's using alternating tabs and spaces. :) 


```txt
        "forceConsistentCasingInFileNames": true,
		"paths": {
			"@/*": ["*"]
		},
```

Reply:
Ah, Justin already commented on that.